### ISSUE TITLE
generateBundle hook should be executed after writing files when calling bundle.write

### ISSUE THREAD

**Issue by egoist** [rollup/rollup#2617]
- Rollup Version: 1.0.0
- Operating System (or Browser): N/A
- Node Version: N/A

### How Do We Reproduce?

Create a simple plugin which uses the `generateBundle` hook to read output file:

```js
const fs = require('fs')

module.exports = () => ({
  generateBundle({ file }, _, isWrite) {
    if (isWrite) {
      fs.readFileSync(file) // Error because it does not exist yet!
    }
  }
})
```

This should be a bug since `onwrite` hook works as expected but it's deprecated in favor of this.

### Expected Behavior

Should be able to access output files when `isWrite` is `true`.

### Actual Behavior

No output files.

It seems `generateBundle` hook is always executed in the `generate` function:

https://github.com/rollup/rollup/blob/4f192335f2cac82897c457ac6cd0d524036ccd5b/src/rollup/index.ts#L293

while `writeOutputFile` is called after that:

https://github.com/rollup/rollup/blob/4f192335f2cac82897c457ac6cd0d524036ccd5b/src/rollup/index.ts#L320

**Comment by ndresx** [comment]
Just stumbled upon the same issue. The `generateBundle` function gets only called once per bundle, with the `isWrite` being `true`. However, the bundle is not available in this moment yet.

### Expected Behavior

The `generateBundle` function should be called twice per bundle and according to the spec, the first time with `isWrite` being `false` =>
> (ongenerate - Use generateBundle - Function hook called when bundle.generate() is being executed.

and the 2nd time with `isWrite` being `true` =>
> onwrite - Use generateBundle - Function hook called when bundle.write() is being executed, after the file has been written to disk.

**Comment by lukastaegert** [comment]
Thanks for the feedback. I agree that there should be a hook that is called *after* the write has been performed. Not sure yet if it should be the same hook (after all it is called "*generate*Bundle").

Just to clarify, what functionality do you want to trigger from this hook? After all, everything that is written is also available via the parameter of this hook. This might help in sketching out how this is addressed. At the moment, my feeling is that I would prefer to add a new hook, e.g. `writeBundle`.

**Comment by egoist** [comment]
I prefer a new hook as well, I use `onwrite` hook in `rollup-plugin-hashbang` to make the output file executable.

**Comment by lukastaegert** [comment]
> I use onwrite hook in rollup-plugin-hashbang to make the output file executable

Ah, that makes sense. I'll see what I can do.

**Comment by lukastaegert** [comment]
Fix at #2617
